### PR TITLE
Feature/file

### PR DESCRIPTION
**Pull request by giann**

### CODE REVIEW

File: src/lauxlib.js [hunk 1]
```diff
@@ -399,6 +400,124 @@ const luaL_newlib = function(L, l) {
     luaL_setfuncs(L, l, 0);
 };
 
+// Only with Node
+if (typeof window === "undefined") {
```
Comment (daurnimator) [inline src/lauxlib.js]: `window` is also undefined in e.g. a web worker.

Comment (giann) [inline src/lauxlib.js]: yes i'll have to come up with a better test

File: src/lauxlib.js [hunk 2]
```diff
@@ -399,6 +400,124 @@ const luaL_newlib = function(L, l) {
...
+            }
+        }
+
+        try {
+            let c;
+            if ((c = skipcomment(lf)))  /* read initial portion */
+                lf.buff[lf.n++] = '\n'.charCodeAt(0);  /* add line to correct line numbers */
+
+            if (c === lua.LUA_SIGNATURE.charCodeAt(0) && filename) {  /* binary file? */
+                // ...
```
Comment (daurnimator) [inline src/lauxlib.js]: something missing? :)

File: src/lauxlib.js [hunk 3]
```diff
@@ -399,6 +400,124 @@ const luaL_newlib = function(L, l) {
...
+
+    const getc = function(lf) {
+        let b = new Buffer(1);
+        let bytes = fs.readSync(lf.f, b, 0, 1, lf.pos);
+        lf.pos += bytes;
+        return bytes > 0 ? b.readUInt8() : null;
+    };
+
+    const skipBOM = function(lf) {
+        let p = "\xEF\xBB\xBF";  /* UTF-8 BOM mark */
```
Comment (daurnimator) [inline src/lauxlib.js]: I'm not sure this will work correctly in JS? ==> Could you add a test that has a UTF-8 BOM?

File: src/ldebug.js [hunk 4]
```diff
@@ -431,7 +431,14 @@ const varinfo = function(L, o) {
             kind = getobjname(ci.func.p, ci.pcOff, stkid - ci.u.l.base);
     }
 
-    return kind ? ` (${kind.funcname} '${kind.name}')` : ``;
+    if (kind) {
+        let funcname = kind.funcname instanceof TValue ? kind.funcname.jsstring() : kind.funcname;
```
Comment (daurnimator) [inline src/ldebug.js]: How could `funcname` ever be a TValue?

File: src/lauxlib.js (hunk 1 above)
Comment (daurnimator) [inline src/lauxlib.js]: Perhaps take a peek at what emscripten does? https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/src/shell.js#L62

Comment (daurnimator) [inline src/lauxlib.js]: Actually, probably better to do a feature check for what we want. i.e. `if (typeof require === "function") { try { fs = require("fs"); } ......`

Comment (giann) [inline src/lauxlib.js]: I'll do that yes

File: src/lauxlib.js [hunk 5]
```diff
@@ -410,9 +410,18 @@ if (typeof window === "undefined") {
             this.f = null;  /* file being read */
             this.buff = new Buffer(1024);  /* area for reading file */
             this.pos = 0;  /* current position in file */
+            this.binary = false;
         }
     }
 
+    const toDataView = function(buffer) {
```
Comment (daurnimator) [inline src/lauxlib.js]: Should make this node.js conditional (as it's particular to node.js's Buffer implementation)

**Pull request merged**